### PR TITLE
COMPAT: catch InvalidIndexError in base Indexer getitem

### LINKED ISSUE #27258: Broken usage of NDFrameIndexer in geopandas
xref https://github.com/pandas-dev/pandas/pull/27223#issuecomment-508828776

The change in that PR broke a test in geopandas. We are using some indexing internals in geopandas that were relying on an InvalidIndexError being catched there (and the PR changed it to only TypeError and KeyError.

A small reproducible example without needing geopandas (but it is dummy code, not doing something useful):

```
from pandas.core.indexing import _NDFrameIndexer

class _CoordinateIndexer(_NDFrameIndexer):

    def _getitem_tuple(self, tup):
        obj = self.obj
        xs, ys = tup
        return obj[xs][ys]

pd.Series._create_indexer('cx', _CoordinateIndexer)

s = pd.Series(range(5))
s.cx[:, :]
```

The below passes on 0.24.2, but gives an error now on master.

The actual code in geopandas is:

https://github.com/geopandas/geopandas/blob/7c66c93f0f76de53615c02810e817ef57fb6a8de/geopandas/base.py#L734-L761

I know this is quite internal code of pandas that is being used here, so we should think about making this cleaner, but short term it would be nice to fix this.

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Closes https://github.com/pandas-dev/pandas/issues/27258

### CODE REVIEW

File: pandas/core/indexing.py [hunk 1]
```diff
@@ -118,7 +118,7 @@ def __getitem__(self, key):
             key = tuple(com.apply_if_callable(x, self.obj) for x in key)
             try:
                 values = self.obj._get_value(*key)
-            except (KeyError, TypeError):
+            except (KeyError, TypeError, InvalidIndexError):
```
Comment (jbrockmendel) [inline pandas/core/indexing.py]: can you add a comment about why this is caught here

Comment (gfyoung) [reply pandas/core/indexing.py]: Do we need to test this somehow as well?

Comment (jorisvandenbossche) [reply pandas/core/indexing.py]: See https://github.com/pandas-dev/pandas/pull/27259 for a discussion on that. I can add a test with the code snippet I show there, but is a rather dummy example ..

Comment (jreback) [reply pandas/core/indexing.py]: I guess this is fine, though why are you not using ``.loc`` and *using* ``__getitem__``?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexing.py (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/core/indexing.py]: > why are you not using `.loc` and using `__getitem__`? 

This is from a custom Indexer attribute in geopandas, so it is on purpose not using `.loc`. Now, I don't say it is the best implementation by using such pandas internals, opened an issue about that in geopandas https://github.com/geopandas/geopandas/issues/1042

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jorisvandenbossche though would avoid using the internals in geopandas :->

**Comment by jbrockmendel** [comment]
I'd still like a comment here about why we are now catching InvalidIndexerError.  Until recently this was catching Exception and I tracked down what actually gets raised, would prefer not to have to do that again in a few months.

**Comment by jorisvandenbossche** [comment]
@jbrockmendel yes, fully agree, was planning to do that so will do a follow up PR

**Pull request merged**